### PR TITLE
Records of Eminence 3 - The Eminencing

### PR DESCRIPTION
**Pull request by Kreidos**
Screenshot - 101020 - 09:55:16 PM
(Relax Zach, 2000 of these lines are just moving records into their own file. :wink: )

Phase 3 mostly adds support for Timed/Daily records, as well as performance improvements/code cleanup.

##### Features: ####
- Timed 4-hour Records (All except magic/physical kills)
- Daily Records (Just Vanquish 30 implemented)
- Add last logout timestamp to chars table (last online time/date)
- Performance improvements + Reduced DB writes (Often significantly)
- Records now kept in separate lua file
- API + code quality Improvements
- Delicious miscellany 

#### Fixes: ####
- Kill credit now counts for party/alliance
- Mob Kill XP now given before Record completion XP (Cases for double-level-ups now in-line with retail behaviour)
- Notification thresholds for all deal/take dmg type records

~~Expect a few follow-up commits if Win32 CI screams bloody murder...~~
Let the record show I have at last submitted a PR which passed CI on the first try. :sunglasses:

**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

**Comment by zach2good** [comment]
image meebs1

**Comment by TeoTwawki** [comment]
@zach2good lien change/add/remove count greatly inflated by github due to breaking out a function that had a fat a-- from one file into its own new file. the 2 files "too large to display" are this.

File: src/map/roe.cpp [hunk 1]
```diff
@@ -43,91 +54,113 @@ void init()
...
     }
     // ShowInfo("\nRoEUtils: %d record entries parsed & available.", RoeBitmaps.ImplementedRecords.count());
     return 0;
 }
 
-bool event(ROE_EVENT eventID, CCharEntity* PChar, RoeDatagramList payload)
+int32 ParseTimedSchedule(lua_State* L)
+{
+    if (lua_isnil(L, -1) || !lua_istable(L, -1))
+            return 0;
```
Comment (TeoTwawki) [inline src/map/roe.cpp]: looks like extra indent snuck in here

Comment (Kreidos) [reply src/map/roe.cpp]: Already caught and fixed in my upcoming chainlink fence commit. :+1:

File: src/map/roe.cpp [hunk 2]
```diff
@@ -21,20 +21,31 @@
...
 
+#define ROE_CACHETIME 15
+
 std::array<RoeCheckHandler, ROE_NONE> RoeHandlers;
 RoeSystemData roeutils::RoeSystem;
 
+void SaveEminenceDataNice(CCharEntity* PChar)
+{
+    if (PChar->m_eminenceCache.lastWriteout < time(nullptr) - ROE_CACHETIME)
+        charutils::SaveEminenceData(PChar);
```
Comment (zach2good) [inline src/map/roe.cpp]: This has been codified in the style guide:
https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#c

`-checks="readability-braces-around-statements"`
```cpp
// Correct ✔️ 
if (x == 5)
{
    function();
}

// Wrong ❌
if (x == 5)
    function();

// Wrong ❌
if (x == 5)
    function(21);
else
{
    function(42);
}
```

File: src/map/roe.cpp [hunk 3]
```diff
@@ -43,91 +54,119 @@ void init()
...
-        ShowError("ROEUtils: Invalid call to RegisterHandler.");
-    }
-
-    // Get Handler Event Type
-    auto event = lua_tointeger(L, 1);
-    if (event == 0 || event >= ROE_NONE)
-    {
-        ShowError("ROEUtils: Unknown Record trigger index %d.", event);
+    if (lua_isnil(L, -1) || !lua_istable(L, -1))
         return 0;
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 4]
```diff
@@ -43,91 +54,119 @@ void init()
...
     }
     // ShowInfo("\nRoEUtils: %d record entries parsed & available.", RoeBitmaps.ImplementedRecords.count());
     return 0;
 }
 
-bool event(ROE_EVENT eventID, CCharEntity* PChar, RoeDatagramList payload)
+int32 ParseTimedSchedule(lua_State* L)
+{
+    if (lua_isnil(L, -1) || !lua_istable(L, -1))
+        return 0;
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 5]
```diff
@@ -244,8 +278,10 @@ bool AddEminenceRecord(CCharEntity* PChar, uint16 recordID)
     // Prevent packet-injection for re-taking completed records which aren't marked repeatable.
     if (roeutils::GetEminenceRecordCompletion(PChar, recordID) && !roeutils::RoeSystem.RepeatableRecords.test(recordID))
         return false;
+    // Prevent packet-injection from taking timed records as normal ones.
+    if (roeutils::RoeSystem.TimedRecords.test(recordID))
+        return false;
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 6]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
             return true;
         }
     }
     return false;
 }
 
+void onCharLoad(CCharEntity* PChar)
+{
+    if (!RoeSystem.RoeEnabled)
+        return;
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 7]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
+
+    // Build eminence lookup map
+    for(int i = 0; i < 31; i++)
+    {
```
Comment (zach2good) [inline src/map/roe.cpp]: `SpaceBeforeParens: ControlStatements`

File: src/map/roe.cpp [hunk 8]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
+
+        {   // 4hr Reset
+            auto* jst = gmtime(&jstnow);
+            jst->tm_hour = jst->tm_hour & 0xFC;
+            jst->tm_min = 0;
+            jst->tm_sec = 0;
+            auto lastJstTimedBlock = timegm(jst) - JST_OFFSET;   // Unix timestamp of the start of the current 4-hr block
+
+            if (lastOnline < lastJstTimedBlock || PChar->m_eminenceLog.active[30] != GetActiveTimedRecord())
+                AddActiveTimedRecord(PChar);
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 9]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
+    {
+        uint16 page = record / 8;
+        uint8 bit = record % 8;
+        PChar->m_eminenceLog.complete[page] &= ~(1 << bit);
+    }
+
+    charutils::SaveEminenceData(PChar);
+
+    for (int i = 0; i < 4; i++)
+        PChar->pushPacket(new CRoeQuestLogPacket(PChar, i));
```
Comment (zach2good) [inline src/map/roe.cpp:473]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 10]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
+}
+
+void CycleTimedRecords()
+{
+    if (!RoeSystem.RoeEnabled)
+        return;
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 11]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
+                AddActiveTimedRecord(PChar);
+            }
+        });
+    });
+}
+
+void CycleDailyRecords()
+{
+    if (!RoeSystem.RoeEnabled)
+        return;
```
Comment (zach2good) [inline src/map/roe.cpp]: `-checks="readability-braces-around-statements"`

File: src/map/roe.cpp [hunk 12]
```diff
@@ -317,11 +353,131 @@ bool SetEminenceRecordProgress(CCharEntity* PChar, uint16 recordID, uint32 progr
...
+
+    zoneutils::ForEachZone([](CZone* PZone){
+        PZone->ForEachChar([](CCharEntity* PChar){
+            ClearDailyRecords(PChar);
+        });
+    });
+}
```
Comment (zach2good) [inline src/map/roe.cpp:505]: `BreakBeforeBraces: Allman` & `Lambdas`

File: src/map/utils/charutils.cpp [hunk 13]
```diff
@@ -437,18 +439,16 @@ namespace charutils
...
-                uint16 record = PChar->m_eminenceLog.active[i];
-                if(record) PChar->m_eminenceCache.activemap.set(record);
-            }
 
             PChar->SetPlayTime(Sql_GetUIntData(SqlHandle, 25));
             PChar->profile.campaign_allegiance = (uint8)Sql_GetIntData(SqlHandle, 26);
             PChar->setStyleLocked(Sql_GetIntData(SqlHandle, 27) == 1 ? true : false);
             PChar->SetMoghancement(Sql_GetUIntData(SqlHandle, 28));
+            PChar->lastOnline = Sql_GetUIntData(SqlHandle, 29);
+
```
Comment (zach2good) [inline src/map/utils/charutils.cpp]: Extra whitespace

File: src/map/utils/charutils.cpp [hunk 14]
```diff
@@ -3726,6 +3726,8 @@ namespace charutils
         {
             //add normal exp
             PChar->jobs.exp[PChar->GetMJob()] += exp;
+            if (PMob != PChar) // Only mob kills count for gain EXP records
+                roeutils::event(ROE_EXPGAIN, PChar, RoeDatagram("exp", exp));
```
Comment (zach2good) [inline src/map/utils/charutils.cpp]: `-checks='readability-braces-around-statements'`

**Review by zach2good — CHANGES_REQUESTED** [review]
Content all looks good, just needs styling according to the style guide.

You're in the enviable position that all your changes are pretty much in files that you're the boss of. So you might be able to set yourself up to do something along the lines of:
```
clang-tidy -header-filter='.*' -checks='readability-braces-around-statements' -fix src/map/roe.cpp 
```

But I haven't tested automatic fixing with clang-tidy yet 🤷‍♂️

File: src/map/roe.cpp (hunk 12 above)
Comment (zach2good) [reply src/map/roe.cpp:505]: Feel free to ignore trying to style these, we're still trying to find a style that doesn't fly in the face of all the other rules, that is actually reproducible by tooling (Lamdas & formatting tooling are not buddies)

**Review by zach2good — APPROVED** [review]

**Comment by Kreidos** [comment]
Oopsie, looks like this merged into release since there was no more `feature/roe` branch to target anymore. Hopefully everything is A-OK!

**Pull request merged**